pandas/core/indexes/base.py

```diff
@@ -2912,7 +2913,22 @@ def union(self, other, sort=None):
         self._assert_can_do_setop(other)
         other, result_name = self._convert_can_do_setop(other)
 
-        if self.dtype != other.dtype:
+        if isinstance(self.dtype, CategoricalDtype) and isinstance(
```

Comment:
This should probably be in `CategoricalIndex._union`

Reply:
If we don't change the current logic, we end up getting `self.dtype != other.dtype` equal True when both are CategoricalIndex, leading to the following line of code where `left` and `right` become Index instead of CategoricalIndex:

```
            dtype = self._find_common_type_compat(other)
            left = self.astype(dtype, copy=False)
            right = other.astype(dtype, copy=False)
            return left.union(right, sort=sort)
```

Dealing with it in `_union`,  would have required to change `union` and `_union` and the current logic. IMHO, I thought it would be better to leave that part of the code intact, since it was doing the right thing for the rest of the cases.